checks/permissions.go

```diff
@@ -0,0 +1,169 @@
+// Copyright 2020 Security Scorecard Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package checks
+
+import (
+	"fmt"
+
+	"github.com/ossf/scorecard/checker"
+	"gopkg.in/yaml.v2"
+)
+
+const CheckPermissions = "Token-Permissions"
+
+//nolint:gochecknoinits
+func init() {
+	registerCheck(CheckPermissions, leastPrivilegedTokens)
+}
+
+func leastPrivilegedTokens(c *checker.CheckRequest) checker.CheckResult {
+	return CheckFilesContent(CheckPermissions, ".github/workflows/*", true, c, validateGitHubActionTokenPermissions)
+}
+
+type GitHubActionWorkflowConfig struct {
+	Permissions *struct {
+		Contents       *string `yaml:"contents"`
```

Comment:
nit: no need for these to be pointers. we can probably simplify things a bit by just using plain `string`s here and checking for `""` instead.

Reply:
Actually thinking about this again, we don't even have to specify all of these explicitly. If we just use a `map[string]string` here we can just iterate over the values here and below without having the list field names.

We probably also need to make `GitHubActionWorkflowConfig` itself a `map[string]interface{}` too and do some runtime type checks to catch the `permissions: read-all|write-all` case (https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#permissions)

Reply:
what are the benefits of not using pointers? I used them because I wanted to differentiate between the field not being present and being set to empty string. (Might not matter in this case, though). But I don't know if there are consequences to using pointers. 

Thanks for the `read-all|write-all` link. I could not find the link yesterday :/.
Is there a way for the yaml parser to understand a union, ie that a field is either a string or a structure? Using a map will work but feels a bit like a hack. What if the structure is more complicated? Is using a map the only way to support unions?